# Stanford Open Policing Project dataset

## Examining the dataset


In [69]:
# Import the pandas library as pd
import pandas as pd

# Read 'police.csv' into a DataFrame named ri
ri = pd.read_csv('police.csv')

# Examine the head of the DataFrame
print(ri.head(1))
print(ri.columns)

              id state   stop_date stop_time location_raw  county_name  \
0  RI-2005-00001    RI  2005-01-02     01:55      Zone K1          NaN   

   county_fips  fine_grained_location police_department driver_gender  \
0          NaN                    NaN               600             M   

     ...     search_conducted  search_type_raw search_type contraband_found  \
0    ...                False              NaN         NaN            False   

  stop_outcome is_arrested stop_duration out_of_state drugs_related_stop  \
0     Citation       False      0-15 Min        False              False   

   district  
0   Zone K1  

[1 rows x 26 columns]
Index(['id', 'state', 'stop_date', 'stop_time', 'location_raw', 'county_name',
       'county_fips', 'fine_grained_location', 'police_department',
       'driver_gender', 'driver_age_raw', 'driver_age', 'driver_race_raw',
       'driver_race', 'violation_raw', 'violation', 'search_conducted',
       'search_type_raw', 'search_type', 'contr

/Users/kavyashreesp/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (8,16) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [70]:
print(ri.isnull())
# Count the number of missing values in each column
print(ri.isnull().sum())

           id  state  stop_date  stop_time  location_raw  county_name  \
0       False  False      False      False         False         True   
1       False  False      False      False         False         True   
2       False  False      False      False         False         True   
3       False  False      False      False         False         True   
4       False  False      False      False         False         True   
5       False  False      False      False         False         True   
6       False  False      False      False         False         True   
7       False  False      False      False         False         True   
8       False  False      False      False         False         True   
9       False  False      False      False         False         True   
10      False  False      False      False         False         True   
11      False  False      False      False         False         True   
12      False  False      False      False         

id                            0
state                         0
stop_date                    10
stop_time                    10
location_raw                  0
county_name              509681
county_fips              509681
fine_grained_location    509681
police_department            10
driver_gender             29097
driver_age_raw            29049
driver_age                30695
driver_race_raw           29073
driver_race               29073
violation_raw             29073
violation                 29073
search_conducted             10
search_type_raw          491919
search_type              491919
contraband_found              0
stop_outcome              29073
is_arrested               29073
stop_duration             29073
out_of_state              29881
drugs_related_stop            0
district                      0
dtype: int64


There are 26 columns and  --rows.

NaN are themissing values.There are many recirds with missing values.

We have to drop those rows  which have NaN values.

Number of missing fields in each column are obtained above using ri.isnull().sum()

### Dropping columns

In [71]:
# Examine the shape of the DataFrame
print(ri.shape)

# Drop the 'county_name' and 'state' columns
ri.drop(['id','county_name', 'state','county_fips','fine_grained_location','driver_age_raw','driver_age','location_raw','police_department','contraband_found','driver_race_raw','search_type_raw','out_of_state'], axis='columns', inplace=True)

# Examine the shape of the DataFrame (again)
print(ri.shape)
print(ri.head(2))

(509681, 26)
(509681, 13)
    stop_date stop_time driver_gender driver_race violation_raw violation  \
0  2005-01-02     01:55             M       White      Speeding  Speeding   
1  2005-01-02     20:30             M       White      Speeding  Speeding   

  search_conducted search_type stop_outcome is_arrested stop_duration  \
0            False         NaN     Citation       False      0-15 Min   
1            False         NaN     Citation       False     16-30 Min   

   drugs_related_stop district  
0               False  Zone K1  
1               False  Zone X4  


### Dropping rows

In [72]:
# Count the number of missing values in each column
print(ri.isnull().sum())
print(ri.shape)
# Drop all rows that are missing 'driver_gender'
ri.dropna(subset=['driver_gender'], inplace=True)

# Count the number of missing values in each column (again)
print(ri.isnull().sum())

# Examine the shape of the DataFrame
print(ri.shape)

stop_date                 10
stop_time                 10
driver_gender          29097
driver_race            29073
violation_raw          29073
violation              29073
search_conducted          10
search_type           491919
stop_outcome           29073
is_arrested            29073
stop_duration          29073
drugs_related_stop         0
district                   0
dtype: int64
(509681, 13)
stop_date                  0
stop_time                  0
driver_gender              0
driver_race                0
violation_raw              0
violation                  0
search_conducted           0
search_type           462822
stop_outcome               0
is_arrested                0
stop_duration              0
drugs_related_stop         0
district                   0
dtype: int64
(480584, 13)


Out of 509681 rows,480584 rows are remaining after dropping missing values in driver_gender

## Examining the data types

In [73]:
ri.dtypes

stop_date             object
stop_time             object
driver_gender         object
driver_race           object
violation_raw         object
violation             object
search_conducted      object
search_type           object
stop_outcome          object
is_arrested           object
stop_duration         object
drugs_related_stop      bool
district              object
dtype: object

In [77]:
ri['is_arrested'] = ri.is_arrested.astype('bool')
ri['search_conducted']=ri.search_conducted.astype('bool')
print(ri.is_arrested.dtype)
print(ri.search_conducted.dtype)

bool
bool


### Creating a DatetimeIndex

In [79]:
#Combining object columns
# Concatenate 'stop_date' and 'stop_time' (separated by a space)
combined = ri.stop_date.str.cat(ri.stop_time,sep=' ')

# Convert 'combined' to datetime format
ri['stop_datetime'] =pd.to_datetime(combined)

# Examine the data types of the DataFrame
print(ri.dtypes)
print(ri.head(2))

stop_date                     object
stop_time                     object
driver_gender                 object
driver_race                   object
violation_raw                 object
violation                     object
search_conducted                bool
search_type                   object
stop_outcome                  object
is_arrested                     bool
stop_duration                 object
drugs_related_stop              bool
district                      object
stop_datetime         datetime64[ns]
dtype: object
    stop_date stop_time driver_gender driver_race violation_raw violation  \
0  2005-01-02     01:55             M       White      Speeding  Speeding   
1  2005-01-02     20:30             M       White      Speeding  Speeding   

   search_conducted search_type stop_outcome  is_arrested stop_duration  \
0             False         NaN     Citation        False      0-15 Min   
1             False         NaN     Citation        False     16-30 Min   

   drugs_r

#### Setting the index

In [80]:
# Set 'stop_datetime' as the index
ri.set_index('stop_datetime', inplace=True)
print(ri.head(2))
# Examine the index
print(ri.index)

# Examine the columns
print(ri.columns)

                      stop_date stop_time driver_gender driver_race  \
stop_datetime                                                         
2005-01-02 01:55:00  2005-01-02     01:55             M       White   
2005-01-02 20:30:00  2005-01-02     20:30             M       White   

                    violation_raw violation  search_conducted search_type  \
stop_datetime                                                               
2005-01-02 01:55:00      Speeding  Speeding             False         NaN   
2005-01-02 20:30:00      Speeding  Speeding             False         NaN   

                    stop_outcome  is_arrested stop_duration  \
stop_datetime                                                 
2005-01-02 01:55:00     Citation        False      0-15 Min   
2005-01-02 20:30:00     Citation        False     16-30 Min   

                     drugs_related_stop district  
stop_datetime                                     
2005-01-02 01:55:00               False  Zone K1  


In [81]:
print(ri.dtypes)

stop_date             object
stop_time             object
driver_gender         object
driver_race           object
violation_raw         object
violation             object
search_conducted        bool
search_type           object
stop_outcome          object
is_arrested             bool
stop_duration         object
drugs_related_stop      bool
district              object
dtype: object


## Questions

##### 1. Do the genders commit different violations?

In [82]:
print(ri.stop_outcome.value_counts())
#method chaining sum()
ri.stop_outcome.value_counts().sum()

Citation            428378
Warning              28840
Arrest Driver        14630
N/D                   3431
No Action             3332
Arrest Passenger      1973
Name: stop_outcome, dtype: int64


480584

In [83]:
#value_counts() suits for categorical data
print(ri.stop_outcome.value_counts(normalize=True))

Citation            0.891370
Warning             0.060010
Arrest Driver       0.030442
N/D                 0.007139
No Action           0.006933
Arrest Passenger    0.004105
Name: stop_outcome, dtype: float64


##### Filtering DataFrame rows

In [85]:
ri.shape
ri.driver_race.value_counts()

White       344716
Black        68577
Hispanic     53123
Asian        12824
Other         1344
Name: driver_race, dtype: int64

In [88]:
white=ri[ri.driver_race=='White']
print(white.shape)

(344716, 13)


In [90]:
asian = ri[ri.driver_race == 'Asian']
asian.stop_outcome.value_counts(normalize=True)

Citation            0.927792
Warning             0.041563
Arrest Driver       0.018013
No Action           0.005848
N/D                 0.004679
Arrest Passenger    0.002105
Name: stop_outcome, dtype: float64

##### Whether male and female drivers tend to commit different types of traffic violations?

In [97]:
print(ri.driver_gender.value_counts())
female=ri[ri.driver_gender=='F']
print(female.shape)
male=ri[ri.driver_gender=='M']
print(male.shape)
print(female.violation.value_counts(normalize=True))
print(male.violation.value_counts(normalize=True))

M    349446
F    131138
Name: driver_gender, dtype: int64
(131138, 13)
(349446, 13)
Citation            0.897337
Warning             0.062636
Arrest Driver       0.020429
No Action           0.007946
N/D                 0.006588
Arrest Passenger    0.005063
Name: stop_outcome, dtype: float64
Citation            0.889130
Warning             0.059025
Arrest Driver       0.034200
N/D                 0.007346
No Action           0.006553
Arrest Passenger    0.003746
Name: stop_outcome, dtype: float64


About two-thirds of female traffic stops are for speeding, whereas stops of males are more balanced among the six categories. This doesn't mean that females speed more often than males, however, since we didn't take into account the number of stops or drivers.

##### Does gender affect who gets a ticket for speeding?

In [100]:
female_and_arrested=ri[(ri.driver_gender == 'F') &
                         (ri.is_arrested == True)]
female_and_arrested.shape

(3343, 13)

In [101]:
female_or_arrested = ri[(ri.driver_gender == 'F') |
                        (ri.is_arrested == True)]
female_or_arrested.shape
(26183, 13)

(26183, 13)

Rules for filtering by multiple conditions

Ampersand (&): only include rows that satisfy both conditions

Pipe (|): include rows that satisfy either condition

Each condition must be surrounded by parentheses

Conditions can check for equality (==), inequality (!=), etc.

Can use more than two conditions

##### When a driver is pulled over for speeding, many people believe that gender has an impact on whether the driver will receive a ticket or a warning.Is that True or Not?

In [104]:
# Create a DataFrame of female drivers stopped for speeding
female_and_speeding = ri[(ri.violation=='Speeding')&(ri.driver_gender=='F')]

# Create a DataFrame of male drivers stopped for speeding
male_and_speeding = ri[(ri.violation=='Speeding')&(ri.driver_gender=='M')]

# Compute the stop outcomes for female drivers (as proportions)
print(female_and_speeding.stop_outcome.value_counts(normalize=True))

# Compute the stop outcomes for male drivers (as proportions)
print(male_and_speeding.stop_outcome.value_counts(normalize=True))
print(female_and_speeding.shape)
print(male_and_speeding.shape)

Citation            0.953247
Warning             0.039003
Arrest Driver       0.005290
Arrest Passenger    0.001033
N/D                 0.000905
No Action           0.000522
Name: stop_outcome, dtype: float64
Citation            0.944636
Warning             0.036086
Arrest Driver       0.015767
Arrest Passenger    0.001265
N/D                 0.001183
No Action           0.001063
Name: stop_outcome, dtype: float64
(86198, 13)
(182538, 13)


Interesting! The numbers are similar for males and females: about 95% of stops for speeding result in a ticket. Thus, the data fails to show that gender has an impact on who gets a ticket for speeding.

##### Does gender affect whose vehicle is searched?

In [105]:
#Mean of Boolean Series represents percentage of True values
import numpy as np
ri.is_arrested.value_counts(normalize=True)

False    0.965452
True     0.034548
Name: is_arrested, dtype: float64

In [106]:
ri.is_arrested.mean()

0.03454755048024903

In [108]:
#Comparing groups using groupby,Studying the arrest rate by police district
ri.district.unique()

array(['Zone K1', 'Zone X4', 'Zone X1', 'Zone K3', 'Zone X3', 'Zone K2'],
      dtype=object)

In [109]:
ri[ri.district == 'Zone K1'].is_arrested.mean()

0.024051181956191714

In [110]:
ri.groupby('district').is_arrested.mean()

district
Zone K1    0.024051
Zone K2    0.028125
Zone K3    0.032020
Zone X1    0.026165
Zone X3    0.033355
Zone X4    0.047291
Name: is_arrested, dtype: float64

`X4` has higher number of arrests

In [112]:
#Grouping by multiple categories
ri.groupby(['district', 'driver_gender']).is_arrested.mean()

district  driver_gender
Zone K1   F                0.015374
          M                0.027779
Zone K2   F                0.019528
          M                0.031619
Zone K3   F                0.022614
          M                0.035451
Zone X1   F                0.020725
          M                0.027561
Zone X3   F                0.026076
          M                0.036446
Zone X4   F                0.038155
          M                0.050246
Name: is_arrested, dtype: float64

In [113]:
ri.groupby(['driver_gender','district']).is_arrested.mean()

driver_gender  district
F              Zone K1     0.015374
               Zone K2     0.019528
               Zone K3     0.022614
               Zone X1     0.020725
               Zone X3     0.026076
               Zone X4     0.038155
M              Zone K1     0.027779
               Zone K2     0.031619
               Zone K3     0.035451
               Zone X1     0.027561
               Zone X3     0.036446
               Zone X4     0.050246
Name: is_arrested, dtype: float64

##### calculate the percentage of all stops that result in a vehicle search, also known as the search rate.

In [120]:
# Check the data type of 'search_conducted'
print(ri.search_conducted.dtype)

# Calculate the search rate by counting the values
print(ri.search_conducted.value_counts(normalize=True))

# Calculate the search rate by taking the mean
print(ri.search_conducted.mean())

bool
False    0.963041
True     0.036959
Name: search_conducted, dtype: float64
0.036959199640437465


the search rate is about 3.6%

In [125]:
#Comparing search rates by gender

# Calculate the search rate for female drivers
print(ri[(ri.driver_gender=='F')].search_conducted.mean())

0.018751239152648355


In [126]:
# Calculate the search rate for each combination of gender and violation
print(ri.groupby(['driver_gender','violation']).search_conducted.mean())

driver_gender  violation          
F              Equipment              0.040245
               Moving violation       0.038021
               Other                  0.045898
               Registration/plates    0.054700
               Seat belt              0.017746
               Speeding               0.007738
M              Equipment              0.070916
               Moving violation       0.059156
               Other                  0.046120
               Registration/plates    0.103589
               Seat belt              0.031705
               Speeding               0.026630
Name: search_conducted, dtype: float64


For all types of violations, the search rate is higher for males than for females, disproving our hypothesis.

##### Does gender affect who is frisked during a search?

In [127]:
#Examining the search types
ri.search_conducted.value_counts()

False    462822
True      17762
Name: search_conducted, dtype: int64

In [128]:
ri.search_type.value_counts(dropna=False)
#value_counts() excludes missing values by default

NaN                                                         462822
Incident to Arrest                                            6998
Probable Cause                                                4989
Reasonable Suspicion                                          1141
Inventory                                                     1101
Protective Frisk                                               879
Incident to Arrest,Inventory                                   649
Incident to Arrest,Probable Cause                              552
Probable Cause,Reasonable Suspicion                            334
Probable Cause,Protective Frisk                                221
Incident to Arrest,Protective Frisk                            158
Incident to Arrest,Inventory,Probable Cause                    151
Inventory,Probable Cause                                       132
Protective Frisk,Reasonable Suspicion                           83
Incident to Arrest,Inventory,Protective Frisk                 

In [129]:
ri.search_type.value_counts()
#dropna=False displays missing values

Incident to Arrest                                          6998
Probable Cause                                              4989
Reasonable Suspicion                                        1141
Inventory                                                   1101
Protective Frisk                                             879
Incident to Arrest,Inventory                                 649
Incident to Arrest,Probable Cause                            552
Probable Cause,Reasonable Suspicion                          334
Probable Cause,Protective Frisk                              221
Incident to Arrest,Protective Frisk                          158
Incident to Arrest,Inventory,Probable Cause                  151
Inventory,Probable Cause                                     132
Protective Frisk,Reasonable Suspicion                         83
Incident to Arrest,Inventory,Protective Frisk                 77
Incident to Arrest,Probable Cause,Protective Frisk            74
Inventory,Protective Fris

In [133]:
#Multiple values are separated by commas
#219 searches in which "Inventory" was the only search type
ri['inventory'] = ri.search_type.str.contains('Inventory', na=False)
#str.contains() returns True if string is found, False if not found
#na=False returns False when it finds a missing value
ri.inventory.dtype
#True means an inventory was done, False means it was not
ri.inventory.sum()
441

441

##### Calculating the inventory rate

In [135]:
ri.inventory.mean()
#0.46% of all traffic stops resulted in an inventory


0.004606894944484211

In [138]:
searched = ri[ri.search_conducted == True]
searched.inventory.mean()
0.13335349259147264
#Correct Answer: 13.3% of searches included an inventory

0.13335349259147264

In [146]:
##Counting protective frisks
# Count the 'search_type' values
print(ri.search_type.value_counts())

# Check if 'search_type' contains the string 'Protective Frisk'
ri['frisk'] = ri.search_type.str.contains('Protective Frisk', na=False)

# Check the data type of 'frisk'
print(ri.frisk.dtype)

# Take the sum of 'frisk'
print(ri.frisk.sum())

Incident to Arrest                                          6998
Probable Cause                                              4989
Reasonable Suspicion                                        1141
Inventory                                                   1101
Protective Frisk                                             879
Incident to Arrest,Inventory                                 649
Incident to Arrest,Probable Cause                            552
Probable Cause,Reasonable Suspicion                          334
Probable Cause,Protective Frisk                              221
Incident to Arrest,Protective Frisk                          158
Incident to Arrest,Inventory,Probable Cause                  151
Inventory,Probable Cause                                     132
Protective Frisk,Reasonable Suspicion                         83
Incident to Arrest,Inventory,Protective Frisk                 77
Incident to Arrest,Probable Cause,Protective Frisk            74
Inventory,Protective Fris

In [157]:
## Comparing frisk rates by gender
# Create a DataFrame of stops in which a search was conducted
searched = ri[(ri.search_conducted==True)]

# Calculate the overall frisk rate by taking the mean of 'frisk'
print(searched.frisk.mean())

# Calculate the frisk rate for each gender
print(searched.groupby(['driver_gender']).frisk.mean())

0.08996734601959239
driver_gender
F    0.066694
M    0.093707
Name: frisk, dtype: float64


##### Does time of day affect arrest rate?